esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6130,329 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+}
+
+bool CWsDfuEx::getGroupNameByClusterNameAndType(IPropertyTree *envTree, const char *clusterName, ClusterType clusterType, StringBuffer &groupName)
+{
+    StringBuffer xpath;
+    if (clusterType == ThorLCRCluster)
+        xpath.set("Software/ThorCluster");
+    else if (clusterType == RoxieCluster)
+        xpath.set("Software/RoxieCluster");
+    else
+        xpath.set("Software/EclAgentProcess");
+
+    Owned<IPropertyTreeIterator> it= envTree->getElements(xpath.str());
+    ForEach(*it)
+    {
+        IPropertyTree &cluster = it->query();
+        if (strieq(clusterName, cluster.queryProp("@name")))
+        {
+            if (clusterType != HThorCluster)
+                getClusterGroupName(cluster, groupName);
+            else
+                groupName.setf("hthor__%s", clusterName);
+            return true;
+        }
+    }
+    return false;
+}
+
+StringBuffer &CWsDfuEx::getGroupNameByClusterName(const char *clusterName, StringBuffer &groupName)
+{
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory(true);
+    Owned<IConstEnvironment> env = factory->openEnvironment();
+    Owned<IPropertyTree> envTree = &env->getPTree();
+    if (getGroupNameByClusterNameAndType(envTree, clusterName, ThorLCRCluster, groupName))
+        return groupName;
+    if (getGroupNameByClusterNameAndType(envTree, clusterName, RoxieCluster, groupName))
+        return groupName;
+    getGroupNameByClusterNameAndType(envTree, clusterName, HThorCluster, groupName);
+    return groupName;
+}
+
+IGroup *CWsDfuEx::queryDFUFileIGroup(const char *clusterName, ClusterType clusterType, const char *clusterTypeEx, StringArray &locations, StringBuffer &groupName)
+{
+    if (!locations.length())
+    {
+        if (isEmptyString(clusterName))
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to get Group: both group location and cluster not defined.");
+
+        getGroupNameByClusterName(clusterName, groupName);
+        if (groupName.isEmpty())
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to get Group for cluster %s.", clusterName);
+
+        GroupType groupType;
+        StringBuffer basedir;
+        Owned<IGroup> group = queryNamedGroupStore().lookup(groupName.str(), basedir, groupType);
+        if (!group)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed to get Group for Group %s.", groupName.str());
+        return group.getClear();
+    }
+
+    StringBuffer locationStr;
+    SocketEndpointArray epa;
+    ForEachItemIn(i, locations)
+    {
+        SocketEndpoint ep(locations.item(i));
+        if (ep.isNull())
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid location '%s'.", locations.item(i));
+        epa.append(ep);
+        locationStr.append(locations.item(i)).append(";");
+    }
+
+    Owned<IGroup> group = createIGroup(epa);
+    groupName.set("wsdfucreate").append(hashc((unsigned char*) locationStr.str(), locationStr.length(), 0));
```

Comment:
"wsdfucreate" would be good if static const at top

Reply:
ok